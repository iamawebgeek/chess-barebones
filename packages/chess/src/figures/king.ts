import {
  AbstractFigure,
  Board8x8,
  Coordinate,
  Direction,
  DirectionToVectorMap,
  getRelativeDirection,
  NotCaptured,
  Player,
  XLine,
} from '@chess-barebones/core';

import { Figure } from './types';

const directions = [
  Direction.EAST,
  Direction.WEST,
  Direction.NORTH,
  Direction.SOUTH,
  Direction.SOUTH_EAST,
  Direction.NORTH_EAST,
  Direction.SOUTH_WEST,
  Direction.NORTH_WEST,
] as const;

export enum Castling {
  SHORT = 'O-O',
  LONG = 'O-O-O',
}

export class King extends AbstractFigure<Figure> {
  public constructor(
    owner: Player,
    board: Board8x8<Figure>,
    coordinate: Coordinate,
  ) {
    super(Figure.KING, owner, board, coordinate);
  }

  public getCastlingFiles(): Record<Castling, XLine> {
    return {
      [Castling.SHORT]: XLine.G,
      [Castling.LONG]: XLine.C,
    };
  }

  private getCastlingMoves() {
    const { coordinate, previousCoordinate, captured } = this.state;
    if (captured || previousCoordinate !== null) {
      return [];
    }
    const castlingFiles = this.getCastlingFiles();
    const castlingRooks = this.getRooksForCastling();
    const moves: {
      rook: AbstractFigure<Figure.ROOK, NotCaptured>;
      type: Castling;
      move: Coordinate;
    }[] = [];
    Object.entries(castlingRooks).forEach(([type, rook]) => {
      if (rook !== null) {
        moves.push({
          rook,
          type: type as Castling,
          move: {
            x: castlingFiles[type as Castling],
            y: coordinate?.y,
          },
        });
      }
    });
    return moves;
  }

  public getRooksForCastling() {
    const castlingRooks: Record<
      Castling,
      null | AbstractFigure<Figure.ROOK, NotCaptured>
    > = {
      [Castling.SHORT]: null,
      [Castling.LONG]: null,
    };

    this.board
      .getPlayerFiguresByName(this.owner, Figure.ROOK)
      .forEach((rook) => {
        const { coordinate, captured, previousCoordinate } = rook.state;
        if (captured || previousCoordinate !== null) {
          return;
        }
        const relativeDirection = getRelativeDirection(
          coordinate,
          this.state.coordinate!,
        );
        if (
          coordinate.y === this.state.coordinate!.y &&
          [Direction.WEST, Direction.EAST].includes(
            relativeDirection as Direction,
          )
        ) {
          castlingRooks[
            relativeDirection === Direction.EAST
              ? Castling.SHORT
              : Castling.LONG
          ] = rook as AbstractFigure<Figure.ROOK, NotCaptured>;
        }
      });
    return castlingRooks;
  }

  public getAllMoves() {
    if (this.state.captured) return [];
    return [
      ...this.getReach(),
      ...this.getCastlingMoves().map(({ move }) => move),
    ];
  }

  public getReach() {
    const { coordinate, captured } = this.state;
    if (captured) {
      return [];
    }
    return directions
      .map((direction) =>
        this.board.getCoordinateWithVector(
          coordinate,
          DirectionToVectorMap[direction],
        ),
      )
      .filter((move) => move !== null);
  }

  public getAvailableMoves() {
    const { coordinate, captured } = this.state;
    if (captured) {
      return [];
    }
    const underCheck =
      this.board
        .getFiguresReachCoordinate(coordinate)
        .filter((figure) => figure.owner !== this.owner).length > 0;
    const castleMoves = underCheck
      ? []
      : this.getCastlingMoves()
          .filter(({ rook, move }) => {
            const rookAt = rook.state.coordinate;
            const lastCoordinateInPath =
              Math.abs(rookAt.x - coordinate.x) >
              Math.abs(move.x - coordinate.x)
                ? this.board.getCoordinateWithVector(
                    rookAt,
                    DirectionToVectorMap[
                      getRelativeDirection(this.state.coordinate!, rookAt)!
                    ],
                  )!
                : move;
            return this.board
              .getPathBetweenCoordinates(coordinate, lastCoordinateInPath)
              .every(
                (coordinate) =>
                  this.board.getFigure(coordinate) === null &&
                  !this.board
                    .getFiguresReachCoordinate(coordinate)
                    .some((figure) => figure.owner !== this.owner),
              );
          })
          .map(({ move }) => move);
    return [
      ...castleMoves,
      ...this.getReach().filter(
        (move) =>
          this.board.getFigure(move)?.owner !== this.owner &&
          !this.board
            .getFiguresReachCoordinate(move)
            .some((figure) => figure !== this && figure.owner !== this.owner),
      ),
    ];
  }
}
