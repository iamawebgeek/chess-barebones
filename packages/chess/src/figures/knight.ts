import {
  AbstractBoard,
  AbstractFigure,
  Coordinate,
  Player,
} from '@chess-barebones/core';

import { Figure } from './types';

const moves = [
  [2, 1],
  [2, -1],
  [-2, 1],
  [-2, -1],
  [1, 2],
  [1, -2],
  [-1, 2],
  [-1, -2],
] as Readonly<[number, number][]>;

export class Knight extends AbstractFigure<Figure> {
  public constructor(
    owner: Player,
    board: AbstractBoard<Figure>,
    coordinate: Coordinate,
  ) {
    super(Figure.KNIGHT, owner, board, coordinate);
  }

  public getAllMoves(): Coordinate[] {
    const { captured, coordinate } = this.state;
    if (captured) return [];
    return moves
      .map((vector) => this.board.getCoordinateWithVector(coordinate, vector))
      .filter((move) => move !== null)
      .flat();
  }

  public getReach(): Coordinate[] {
    return this.getAllMoves();
  }

  public getAvailableMoves(): Coordinate[] {
    return this.getReach().filter(
      (coordinate) => this.board.getFigure(coordinate)?.owner !== this.owner,
    );
  }
}
