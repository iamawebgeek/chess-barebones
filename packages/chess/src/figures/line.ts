import {
  AbstractFigure,
  Coordinate,
  Direction,
  DirectionToVectorMap,
} from '@chess-barebones/core';

import { Figure } from './types';

export abstract class LineMovingFigure<
  T extends string,
> extends AbstractFigure<T> {
  public abstract getDirections(): Direction[];

  public getAllMoves() {
    const { captured, coordinate } = this.state;
    if (captured) return [];
    return this.getDirections()
      .map((direction) => {
        let next = this.board.getCoordinateWithVector(
          coordinate,
          DirectionToVectorMap[direction],
        );
        const moves: Coordinate[] = [];
        while (next !== null) {
          moves.push(next);
          next = this.board.getCoordinateWithVector(
            next,
            DirectionToVectorMap[direction],
          );
        }
        return moves;
      })
      .flat();
  }

  public getReach() {
    const { captured, coordinate } = this.state;
    if (captured) return [];
    return this.getDirections()
      .map((direction) => {
        let next = this.board.getCoordinateWithVector(
          coordinate,
          DirectionToVectorMap[direction],
        );
        const moves: Coordinate[] = [];
        while (next !== null) {
          const figureAtCoordinate = this.board.getFigure(next);
          const owner = figureAtCoordinate?.owner ?? null;
          moves.push(next);
          next = this.board.getCoordinateWithVector(
            next,
            DirectionToVectorMap[direction],
          );
          if (
            owner !== null &&
            (figureAtCoordinate.name !== Figure.KING ||
              figureAtCoordinate.owner === this.owner)
          ) {
            break;
          }
        }
        return moves;
      })
      .flat();
  }

  public getAvailableMoves() {
    const { captured, coordinate } = this.state;
    if (captured) return [];
    return this.getDirections()
      .map((direction) => {
        let next = this.board.getCoordinateWithVector(
          coordinate,
          DirectionToVectorMap[direction],
        );
        const moves: Coordinate[] = [];
        while (next !== null) {
          const figureAtCoordinate = this.board.getFigure(next);
          const owner = figureAtCoordinate?.owner ?? null;
          moves.push(next);
          next = this.board.getCoordinateWithVector(
            next,
            DirectionToVectorMap[direction],
          );
          if (owner !== null) {
            if (owner === this.owner) {
              moves.pop();
            }
            break;
          }
        }
        return moves;
      })
      .flat();
  }
}
