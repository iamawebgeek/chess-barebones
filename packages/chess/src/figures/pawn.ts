import {
  AbstractBoard,
  AbstractFigure,
  Coordinate,
  Direction,
  DirectionToVectorMap,
  Move,
  Player,
} from '@chess-barebones/core';

import { Figure } from './types';

const INITIAL_LOCATION_TO_ATTACK_DIRECTION = {
  [Direction.NORTH]: [Direction.NORTH_WEST, Direction.NORTH_EAST],
  [Direction.SOUTH]: [Direction.SOUTH_WEST, Direction.SOUTH_EAST],
};

export const PAWN_PROMOTION = Symbol('PAWN_PROMOTION');

export class Pawn extends AbstractFigure<Figure> {
  public constructor(
    owner: Player,
    board: AbstractBoard<Figure>,
    coordinate: Coordinate,
  ) {
    super(Figure.PAWN, owner, board, coordinate);
  }

  private get headedDirection() {
    return this.owner.flank === Direction.NORTH
      ? Direction.SOUTH
      : Direction.NORTH;
  }

  private checkForPromotion(coordinate: Coordinate) {
    const nextCell = this.board.getCoordinateWithVector(
      coordinate,
      DirectionToVectorMap[this.headedDirection],
    );
    return nextCell === null;
  }

  public move(to: Coordinate | string) {
    if (typeof to === 'string') {
      to = this.board.deserializeCoordinate(to);
    }
    // en-passant capture
    if (
      to.x !== this.state.coordinate!.x &&
      this.board.getFigure(to) === null
    ) {
      this.board.lastMoved?.captureBy(this.owner);
    }
    super.move(to);
  }

  public getAllMoves() {
    const { captured, coordinate } = this.state;
    if (captured) return [];
    const nextCoordinates = [
      this.board.getCoordinateWithVector(
        coordinate,
        DirectionToVectorMap[this.headedDirection],
      ),
    ] as Move[];
    if (nextCoordinates[0]) {
      const nextCell = this.board.getCoordinateWithVector(
        nextCoordinates[0],
        DirectionToVectorMap[this.headedDirection],
      );
      if (this.checkForPromotion(nextCoordinates[0])) {
        nextCoordinates[0].requiresInput = PAWN_PROMOTION;
      }
      if (this.state.previousCoordinate === null && nextCell !== null) {
        nextCoordinates.push(nextCell);
      }
    }
    return [
      ...this.getReach(),
      ...(nextCoordinates[0] ? nextCoordinates : []),
    ] as Coordinate[];
  }

  public getReach() {
    const { captured, coordinate } = this.state;
    if (captured) return [];
    return INITIAL_LOCATION_TO_ATTACK_DIRECTION[this.headedDirection]
      .map((direction) =>
        this.board.getCoordinateWithVector(
          coordinate,
          DirectionToVectorMap[direction],
        ),
      )
      .filter((move) => move !== null);
  }

  public getAvailableMoves() {
    const { captured, coordinate } = this.state;
    if (captured) return [];
    const reach = this.getReach();
    const otherMoves = this.getAllMoves().slice(reach.length);
    return [
      ...reach
        .filter((move) => {
          // en-passant
          const lastFigure = this.board.lastMoved;
          if (lastFigure !== null && lastFigure.name === Figure.PAWN) {
            const lastState = lastFigure.state;
            const prev = lastState.previousCoordinate;
            const curr = lastState.coordinate;
            // Must be immediate previous move, adjacent file target, same rank as our pawn,
            // enemy pawn must have moved exactly two squares last move
            if (
              prev &&
              curr &&
              lastFigure.owner !== this.owner &&
              curr.y === coordinate.y &&
              curr.x === move.x &&
              Math.abs(curr.y - prev.y) === 2
            ) {
              return true;
            }
          }
          const owner = this.board.getFigure(move)?.owner;
          return owner && owner !== this.owner;
        })
        .map((coordinate) =>
          this.checkForPromotion(coordinate)
            ? ({ ...coordinate, requiresInput: PAWN_PROMOTION } as Move)
            : coordinate,
        ),
      ...(this.board.getFigure(otherMoves[0]) === null
        ? otherMoves.filter(
            (coordinate) => this.board.getFigure(coordinate) === null,
          )
        : []),
    ];
  }
}
