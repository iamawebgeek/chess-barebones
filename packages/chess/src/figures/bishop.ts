import {
  AbstractBoard,
  Coordinate,
  Direction,
  Player,
} from '@chess-barebones/core';

import { LineMovingFigure } from './line';
import { Figure } from './types';

export class Bishop extends LineMovingFigure<Figure> {
  private readonly directions = [
    Direction.SOUTH_EAST,
    Direction.NORTH_EAST,
    Direction.SOUTH_WEST,
    Direction.NORTH_WEST,
  ];

  public constructor(
    owner: Player,
    board: AbstractBoard<Figure>,
    coordinate: Coordinate,
  ) {
    super(Figure.BISHOP, owner, board, coordinate);
  }

  public getDirections(): Direction[] {
    return this.directions;
  }
}
