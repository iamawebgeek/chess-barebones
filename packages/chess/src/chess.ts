import {
  AbstractBoard,
  combineHandlers,
  Direction,
  Game,
  Handler,
  InvalidStateError,
  Player,
} from '@chess-barebones/core';

import { Figure } from './figures';
import {
  CheckmateProcessor,
  FiftyMovesProcessor,
  InsufficientMaterialsProcessor,
  RepetitionProcessor,
  StalemateProcessor,
} from './processors';
import { AlgebraicNotationSerializer, RegularChessMove } from './serializers';

export enum Color {
  WHITE = 'white',
  BLACK = 'black',
}

export class RegularChess extends Game<Figure, RegularChessMove, Color> {
  public constructor(board: AbstractBoard<Figure>, handler: Handler = {}) {
    const serializer = new AlgebraicNotationSerializer(board);
    super(
      board,
      serializer,
      combineHandlers([
        handler,
        // Do not change
        // The order is important
        new CheckmateProcessor(board),
        new StalemateProcessor(board),
        new InsufficientMaterialsProcessor(board),
        new FiftyMovesProcessor(board),
        new RepetitionProcessor(board),
      ]),
    );
    this.addPlayer(new Player(Color.WHITE, Direction.SOUTH));
    this.addPlayer(new Player(Color.BLACK, Direction.NORTH));
  }

  public start() {
    if (this.state.players.length !== 2) {
      throw new InvalidStateError(
        this,
        'Regular chess can be played with only 2 players, consider making your variation with Game class',
      );
    }
    super.start();
  }

  public getPlayerByColor(color: Color) {
    return this.state.players.find((player) => player.color === color)!;
  }
}
