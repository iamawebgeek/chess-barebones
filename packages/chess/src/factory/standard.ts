import {
  AbstractBoard,
  AbstractFigure,
  Coordinate,
  FigureFactory,
  NotCaptured,
  Player,
} from '@chess-barebones/core';

import { applyCheckDecorator, applyPinDecorator } from '../decorators';
import { Bishop, Figure, King, Knight, Pawn, Queen, Rook } from '../figures';

const DecoratedFigures = {
  [Figure.KING]: King,
  [Figure.BISHOP]: applyCheckDecorator(applyPinDecorator(Bishop)),
  [Figure.ROOK]: applyCheckDecorator(applyPinDecorator(Rook)),
  [Figure.KNIGHT]: applyCheckDecorator(applyPinDecorator(Knight)),
  [Figure.QUEEN]: applyCheckDecorator(applyPinDecorator(Queen)),
  [Figure.PAWN]: applyCheckDecorator(applyPinDecorator(Pawn)),
};

export class ChessFigureFactory implements FigureFactory<Figure> {
  public create(
    name: Figure,
    player: Player,
    board: AbstractBoard<Figure>,
    coordinate: Coordinate,
  ) {
    const DecoratedFigure = DecoratedFigures[name];
    return new DecoratedFigure(player, board, coordinate) as AbstractFigure<
      Figure,
      NotCaptured
    >;
  }
}
