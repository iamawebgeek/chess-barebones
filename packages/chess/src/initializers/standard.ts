import { Board8x8, Handler, Player, XLine, YLine } from '@chess-barebones/core';

import { Figure } from '../figures';

const frontlinePieces = [
  [
    Figure.PAWN,
    [XLine.A, XLine.B, XLine.C, XLine.D, XLine.E, XLine.F, XLine.G, XLine.H],
  ],
] as const;
const backlinePieces = [
  [Figure.ROOK, [XLine.A, XLine.H]],
  [Figure.KNIGHT, [XLine.B, XLine.G]],
  [Figure.BISHOP, [XLine.C, XLine.F]],
  [Figure.QUEEN, [XLine.D]],
  [Figure.KING, [XLine.E]],
] as const;

export class StandardChessInitializer implements Handler {
  public constructor(private readonly board: Board8x8<Figure>) {}

  public onPlayerAdded(player: Player) {
    const [frontline, backline] =
      player.color === 'white' ? [YLine._2, YLine._1] : [YLine._7, YLine._8];
    frontlinePieces.forEach(([figureName, xLines]) => {
      xLines.forEach((x) => {
        this.board.createFigure(figureName, player, { x, y: frontline });
      });
    });
    backlinePieces.forEach(([figureName, xLines]) => {
      xLines.forEach((x) => {
        this.board.createFigure(figureName, player, { x, y: backline });
      });
    });
  }
}
