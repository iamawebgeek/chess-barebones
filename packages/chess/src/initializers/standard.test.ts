import { Board8x8, Direction, Player } from '@chess-barebones/core';
import { describe, expect, it } from 'vitest';

import { StandardChessInitializer } from './standard';
import { Color } from '../chess';
import { ChessFigureFactory } from '../factory';
import { Figure } from '../figures';

describe('StandardChessInitializer', () => {
  it('creates same figures', () => {
    const board = new Board8x8<Figure>(new ChessFigureFactory());
    const initializer = new StandardChessInitializer(board);
    const player1 = new Player(Color.WHITE, Direction.SOUTH);
    const player2 = new Player(Color.BLACK, Direction.NORTH);
    initializer.onPlayerAdded(player1);
    initializer.onPlayerAdded(player2);
    expect(board.serializePosition()).toMatchSnapshot();
  });
});
