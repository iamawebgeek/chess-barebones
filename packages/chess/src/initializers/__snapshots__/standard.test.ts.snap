// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StandardChessInitializer > creates same figures 1`] = `"black-bishop-1:c8:-;black-bishop-2:f8:-;black-king-1:e8:-;black-knight-1:b8:-;black-knight-2:g8:-;black-pawn-1:a7:-;black-pawn-2:b7:-;black-pawn-3:c7:-;black-pawn-4:d7:-;black-pawn-5:e7:-;black-pawn-6:f7:-;black-pawn-7:g7:-;black-pawn-8:h7:-;black-queen-1:d8:-;black-rook-1:a8:-;black-rook-2:h8:-;white-bishop-1:c1:-;white-bishop-2:f1:-;white-king-1:e1:-;white-knight-1:b1:-;white-knight-2:g1:-;white-pawn-1:a2:-;white-pawn-2:b2:-;white-pawn-3:c2:-;white-pawn-4:d2:-;white-pawn-5:e2:-;white-pawn-6:f2:-;white-pawn-7:g2:-;white-pawn-8:h2:-;white-queen-1:d1:-;white-rook-1:a1:-;white-rook-2:h1:-"`;
